cmds/sshd/sshd.go

```diff
@@ -0,0 +1,238 @@
+	}
+	go copy(p.Ptm, c)
+	go copy(c, p.Ptm)
+	return nil
+}
+
+func newPTY(b []byte) (*pty.Pty, error) {
+	ptyReq := &ptyReq{}
+	err := ssh.Unmarshal(b, ptyReq)
+	dprintf("newPTY: %q", ptyReq)
+	if err != nil {
+		return nil, err
+	}
+	p, err := pty.New()
+	ws, err := p.TTY.GetWinSize()
+	if err != nil {
+		return nil, err
+	}
+	ws.Row = uint16(ptyReq.Row)
+	ws.Ypixel = uint16(ptyReq.Ypixel)
+	ws.Col = uint16(ptyReq.Col)
+	ws.Xpixel = uint16(ptyReq.Xpixel)
+	dprintf("newPTY: Set winsizes to %v", ws)
+	if err := p.TTY.SetWinSize(ws); err != nil {
+		return nil, err
+	}
+	dprintf("newPTY: set TERM to %q", ptyReq.TERM)
+	if err := os.Setenv("TERM", ptyReq.TERM); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func init() {
+	for _, s := range shells {
+		if _, err := exec.LookPath(s); err == nil {
+			shell = s
+		}
+	}
+}
+
+func session(chans <-chan ssh.NewChannel) {
+	var p *pty.Pty
+	// Service the incoming Channel channel.
+	for newChannel := range chans {
+		// Channels have a type, depending on the application level
+		// protocol intended. In the case of a shell, the type is
+		// "session" and ServerShell may be used to present a simple
+		// terminal interface.
+		if newChannel.ChannelType() != "session" {
+			newChannel.Reject(ssh.UnknownChannelType, "unknown channel type")
+			continue
+		}
+		channel, requests, err := newChannel.Accept()
+		if err != nil {
+			log.Fatalf("Could not accept channel: %v", err)
+		}
+
+		// Sessions have out-of-band requests such as "shell",
+		// "pty-req" and "env".  Here we handle only the
+		// "shell" request.
+		go func(in <-chan *ssh.Request) {
+			for req := range in {
+				dprintf("Request %v", req.Type)
+				switch req.Type {
+				case "shell":
+					if p == nil {
+						p, err = pty.New()
+						if err != nil {
+							log.Printf("sshd: pty.New failed, not running a shell")
+							break
+						}
+					}
+					err := runShell(channel, p, shell)
+					req.Reply(true, []byte(fmt.Sprintf("%v", err)))
+				case "pty-req":
+					p, err = newPTY(req.Payload)
+					req.Reply(err == nil, nil)
+				default:
+					fmt.Printf("Not handling req %v %q", req, string(req.Payload))
+					req.Reply(false, nil)
+				}
+			}
+		}(requests)
+
+	}
+}
+
+func main() {
+	flag.Parse()
+	if *debug {
+		dprintf = log.Printf
+	}
+	// Public key authentication is done by comparing
+	// the public key of a received connection
+	// with the entries in the authorized_keys file.
+	authorizedKeysBytes, err := ioutil.ReadFile(*keys)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	authorizedKeysMap := map[string]bool{}
+	for len(authorizedKeysBytes) > 0 {
+		pubKey, _, _, rest, err := ssh.ParseAuthorizedKey(authorizedKeysBytes)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		authorizedKeysMap[string(pubKey.Marshal())] = true
+		authorizedKeysBytes = rest
+	}
+
+	// An SSH server is represented by a ServerConfig, which holds
+	// certificate details and handles authentication of ServerConns.
+	config := &ssh.ServerConfig{
```

Comment:
Nit, don't take the pointer of ssh.ServerConfig during assignment. Do it on line 227.

Reply:
I'm not so sure. There are 2500 instances of this idiom in the go source tree. Why is it a problem here?

Reply:
It improves readability. If you look at line 229 in isolation:

```
conn, chans, reqs, err := ssh.NewServerConn(nConn, config)
```

What can we say about the config variable? Currently only its name, thus we need to scan back up to line 186. 

With:

```
conn, chans, reqs, err := ssh.NewServerConn(nConn, &config)
```

We have additional context. We know we're sharing the config variable with ssh.NewServerConn. We know config is going out on heap.

In this specific instance it's a nit, since all the code is within 10s of lines of each other. If you happen to have a Safari Books subscription, checkout William Kennedy's Ultimate Go Programming video course, 2.3.3 Pointers, Part 3 at around 7mins in.

Reply:
Also not sure how you're finding 2500 instances of the `ptr = &var` idiom, but I can only find ~650 in the Go source tree. With ~3159 total count of times a pointer is taken of something:

```
[go]$ find . -name "*.go" -print0 | xargs -0 grep "\&[:alpha:]" | grep -v \; | wc -l
    3159
[go]$ find . -name "*.go" -print0 | xargs -0 grep "= \&[:alpha:]" | grep -v \; | wc -l
     650
```

Reply:
yeah good point, I counted wrong.. I still don't agree however. I kinda go with what the Go authors do. My standard reference is go/src.

And people should not be consciously thinking about things going out on heap! I hope he's not using that to support his argument.

Reply:
I disagree on that people shouldn't be thinking about what goes on the heap. The heap is GC'ed. A stack is swept clean on initialization and thus doesn't need require a GC.